Don't override message intent and reply to address if the headers contain value

indualagarsamy opened the pull request:
Fixes https://github.com/Particular/NServiceBus.Transport.Msmq/issues/17

cc / @timbussmann 
@Particular/msmq-transport-maintainers

andreasohlund commented on src/NServiceBus.Transport.Msmq.Tests/MsmqUtilitiesTests.cs:
```diff
@@ -61,6 +61,54 @@ public void Should_fetch_the_replyToAddress_from_responsequeue_for_backwards_com
+            message.ResponseQueue = new MessageQueue(new MsmqAddress("local", RuntimeEnvironment.MachineName).FullPath);
+            var headers = MsmqUtilities.ExtractHeaders(message);
+
+            Assert.AreNotEqual("local@" + RuntimeEnvironment.MachineName, headers[Headers.ReplyToAddress]);
```
This seems redundant? (the next assert covers this?)

indualagarsamy replied:
yeah, I'll remove the redundant assert.

andreasohlund commented on src/NServiceBus.Transport.Msmq.Tests/MsmqUtilitiesTests.cs:
```diff
@@ -61,6 +61,54 @@ public void Should_fetch_the_replyToAddress_from_responsequeue_for_backwards_com
+            message.AppSpecific = 3; //Send = 1, Publish = 2, Subscribe = 3, Unsubscribe = 4 and Reply = 5 
+            var headers = MsmqUtilities.ExtractHeaders(message);
+
+            Assert.AreNotEqual("Subscribe", headers[Headers.MessageIntent]);
```
Same here, redundant?

indualagarsamy replied:
yeah, I'll remove the redundant assert.

indualagarsamy commented:
@andreasohlund - I have fixed the failing unit test problem. The original code only assigned the header value if there was a proper value, whereas I was setting the value always which is incorrect. I have fixed it. Please have a look now. I have also squashed my commits and rebased on the release branch to make review easier.

andreasohlund commented on src/NServiceBus.Transport.Msmq/MsmqUtilities.cs:
```diff
@@ -47,28 +47,25 @@ static MsmqAddress GetIndependentAddressForQueue(MessageQueue q)
+            
             //note: we can drop this line when we no longer support interop btw v3 + v4
-            if (msmqMessage.ResponseQueue != null)
+            if (msmqMessage.ResponseQueue != null && headers.TryGetValue(Headers.ReplyToAddress, out var _) == false)
```
Since you don't care about the value I think `!headers.ContainsKey(...)` is more correct?

andreasohlund commented on src/NServiceBus.Transport.Msmq/MsmqUtilities.cs:
```diff
@@ -47,28 +47,25 @@ static MsmqAddress GetIndependentAddressForQueue(MessageQueue q)
             }
 
-            if (Enum.IsDefined(typeof(MessageIntentEnum), msmqMessage.AppSpecific))
+            if (Enum.IsDefined(typeof(MessageIntentEnum), msmqMessage.AppSpecific) && headers.TryGetValue(Headers.MessageIntent, out var _) == false)
```
Since you don't care about the value I think `!headers.ContainsKey(...)` is more correct?

indualagarsamy commented:
@andreasohlund - Checking against ContainsKey is much better. I've fixed the code per your suggestion.

andreasohlund commented on src/NServiceBus.Transport.Msmq/MsmqUtilities.cs:
```diff
@@ -47,28 +47,25 @@ static MsmqAddress GetIndependentAddressForQueue(MessageQueue q)
+            
             //note: we can drop this line when we no longer support interop btw v3 + v4
-            if (msmqMessage.ResponseQueue != null)
+            if (msmqMessage.ResponseQueue != null && headers.ContainsKey(Headers.ReplyToAddress) == false)
```
More nitpicking: 

`headers.ContainsKey(Headers.ReplyToAddress) == false`

 should be

`!headers.ContainsKey(Headers.ReplyToAddress)`

andreasohlund commented on src/NServiceBus.Transport.Msmq/MsmqUtilities.cs:
```diff
@@ -47,28 +47,25 @@ static MsmqAddress GetIndependentAddressForQueue(MessageQueue q)
             }
 
-            if (Enum.IsDefined(typeof(MessageIntentEnum), msmqMessage.AppSpecific))
+            if (Enum.IsDefined(typeof(MessageIntentEnum), msmqMessage.AppSpecific) && headers.ContainsKey(Headers.MessageIntent) == false)
```
Same as above

indualagarsamy commented:
@andreasohlund - fixed now.
